rust/dialog-storage/src/storage/backend/s3/access.rs

```diff
@@ -0,0 +1,705 @@
+
+        // Build final URL with all query parameters (original + signing + signature)
+        let mut url = request.url().clone();
+        url.set_query(None); // Clear existing query params (we'll add them all back)
+        {
+            let mut query = url.query_pairs_mut();
+            for (k, v) in &query_params {
+                query.append_pair(k, v);
+            }
+            query.append_pair("X-Amz-Signature", &signature.to_string());
+        }
+
+        Ok(Authorized { url, headers })
+    }
+}
+
+/// Trait for request invocations that can be authorized and sent.
+///
+/// Implementors describe the HTTP method, URL, and optional checksum/ACL
+/// for S3 requests. The [`Request`](super::Request) trait extends this
+/// with a body and the ability to perform the request.
+pub trait Invocation {
+    /// The HTTP method for this request.
+    fn method(&self) -> &'static str;
+
+    /// The URL for this request.
+    fn url(&self) -> &Url;
+
+    /// The checksum of the body, if any.
+    fn checksum(&self) -> Option<&Checksum> {
+        None
+    }
+
+    /// The ACL for this request, if any.
+    fn acl(&self) -> Option<Acl> {
+        None
+    }
+}
+
+/// An authorized request ready to be sent.
+#[derive(Debug)]
+pub struct Authorized {
+    /// The presigned URL
+    pub url: Url,
+    /// Headers that must be included in the HTTP request
+    pub headers: Vec<(String, String)>,
+}
+
+/// HMAC-SHA256 signature bytes.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct Signature(Vec<u8>);
+
+impl std::fmt::Display for Signature {
+    /// Displays hex encoded representation of the signature
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&hex_encode(&self.0))
+    }
+}
+
+/// Errors that can occur during signing.
+#[derive(Debug)]
+pub enum SigningError {
+    /// The endpoint URL is invalid (e.g., missing host).
+    InvalidEndpoint(String),
+    /// Failed to parse a URL.
+    UrlParse(url::ParseError),
+}
+
+impl std::fmt::Display for SigningError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::InvalidEndpoint(msg) => write!(f, "invalid endpoint: {}", msg),
+            Self::UrlParse(e) => write!(f, "URL parse error: {}", e),
+        }
+    }
+}
+
+impl std::error::Error for SigningError {}
+
+impl From<url::ParseError> for SigningError {
+    fn from(e: url::ParseError) -> Self {
+        Self::UrlParse(e)
+    }
+}
+
+// ============================================================================
+// Helper Functions
+// ============================================================================
+
+fn current_time() -> DateTime<Utc> {
+    #[cfg(target_arch = "wasm32")]
+    {
+        DateTime::<Utc>::from(SystemTime::now().to_std())
+    }
+    #[cfg(not(target_arch = "wasm32"))]
+    {
+        DateTime::<Utc>::from(SystemTime::now())
+    }
+}
+
+fn hex_encode(bytes: &[u8]) -> String {
```

Comment:
Some human-readable documentation would be nice even though this is private API.